### PR TITLE
Bianca & Mariana - Pipes - Scrabble 

### PR DESCRIPTION
**Pull request by biciclista22**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |  We don't want to create a new instance variable for a word every time we call the score method.   |
|   Describe an example of a test you wrote for a _nominal case_. |  An example of a nominal case test is for our tiles remaining method in tilebag sepc. We called the draw tiles method and input the number of tiles we wanted to remove. We confirmed the number of tiles remaining in the bag. We pulled 3 and expected 95 since the bag of tiles has 98. 	|
|   Describe an example of a test you wrote for an _edge case_. | In the player spec, we wanted to test what would happen for the method highest_scoring_word_score in the Player method if the player hasn't played yet (and thus the array is empty). We included an argument error in the method and tried to see if it would be called in our test.  	|
|   What was your final test coverage percentage?	|  Our final test coverage was about 99.29% if we don't include our dictionary method and tests. 	|
|   Is there a specific piece of code you'd like feedback on?	|   We would love some feedback on our Player class. It is one of the more complex class methods we created and think it could likely be DRY'ed and refactored. 	|

### CODE REVIEW

File: specs/player_spec.rb [hunk 1]
```diff
@@ -0,0 +1,126 @@
+require_relative 'spec_helper.rb'
+
+describe "Class player" do
+  before do
+    @tile_bag = Scrabble::TileBag.new
+  end
+  describe "Initialize" do
+    it "Return to name" do
+      Scrabble::Player.must_respond_to :name
```
Comment (Hamled) [inline specs/player_spec.rb:9]: This isn't doing what you think. Here we're checking if the `Player` class itself (_not_ an instance of it) responds to the method `name`.

As it happens this test passes because every class has a "built-in" class method called `name` which just returns the name of the class:
```ruby
class Foo
end

Foo.name #=> "Foo"
```

For this test to reflect what you're actually trying to test we would need something like:
```ruby
Scrabble::Player.new("test").must_respond_to :name
```

however that behavior is already covered in the following test so I think we can safely remove this test entirely.

**Comment by Hamled** [comment]
# Scrabble
## What We're Looking For

|  Feature  |   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions  | :white_check_mark: |
|   Both Teammates contributed to the codebase	| :white_check_mark:	|
|   Regular Commits with meaningful commit messages  | :white_check_mark:	|
|  **Scoring class**	|   	|
|  `score` method	| :white_check_mark: |
|   Uses appropriate data structure to store the letter score	| :white_check_mark:	|
|   Appropriately handles edge cases	| :white_check_mark: |
|   Tests `score` edge cases (one-letter word, 7 letter word)	| :white_check_mark: |
|  `highest_scoring_word` method	| :white_check_mark:	|
|   Appropriately handles edge cases	| :white_check_mark:	|
|   Tests `highest_scoring_word` edge cases (ties, no words, no array)	| Partial. There are tests for the two tie-based edge cases (7-letter word or shortest word), but no tests for when invalid data is provided. We should also have tests for when the method is called with `[]` or with a non-array parameter like `"words"`. |
|  **Player class**	|   	|
|  Uses an array to keep track of words played	| :white_check_mark: |
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	| No. We seem to have a single method, `highest_scoring_word_score` which is returning both the highest scoring word and the score for that word. However, that method is not making use of the `Scoring.highscore_from` method to DRY up some of the logic. |
|  Returns `true` or `false` from `won?` method | :white_check_mark: However we could simplify the code to accomplish this. See my additional code comments. |
|  Tests edge cases on the `play(word)` method for words that cannot be scored | :white_check_mark: However there are some bugs in the `delete_tiles_from_tiles_array` method that would be exposed if we created some additional tests. |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	| :white_check_mark: |
|  `draw_tiles` method uses the parameter to draw a set of random tiles | :white_check_mark: |
| `tiles_remaining` method returns a count	| :white_check_mark: |
|  Tests `draw_tiles` edge cases for different parameter inputs	| Partial. We could have a test for attempting to draw zero tiles, which should probably raise ArgumentError similarly to when attempting to draw too many. We could also have a test that `draw_tiles` only returns letters which are definitely left in the bag. |

File: specs/scoring_spec.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  end#Describe Score class methos
+
+  describe "High Score method" do
+    # it "Returns array an array for array of words that is not empty " do
+    #   array_words = ["house", "mama", "florida", "flan", "dulce"]
+    #   Scrabble::Scoring.highscore_from(array_words).must_be_instance_of Array
+    #   Scrabble::Scoring.highscore_from(array_words).wont_be_empty
+    # end
+
+    it "Score return the correct score for each word in the array" do
```
Comment (Hamled) [inline specs/scoring_spec.rb:31]: We should probably remove these two commented out tests.

File: specs/scoring_spec.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    it "Retunrs 0 for an empty string" do
+      Scrabble::Scoring.score("").must_equal 0
+    end
+
+    it "Returns the score for a word" do
+      Scrabble::Scoring.score("CASA").must_equal 6
+    end
+
+    it "Returns the score even when lowercase word" do
+      Scrabble::Scoring.score('CASA').must_equal 6
```
Comment (Hamled) [inline specs/scoring_spec.rb:16]: This word is not lowercase as the test title implies it should be.

File: specs/player_spec.rb [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    end
+
+    it "Player instance variable is empty upon initialization" do
+      Scrabble::Player.new("Mariana").plays.must_equal []
+    end
+
+    describe "play method" do
+      before do
+        @new_player = Scrabble::Player.new("Ecuador", true)
+        @new_player.draw_tiles(@tile_bag, ["P", "E", "R", "E", "Z", "A", "M", "M", "A"])
```
Comment (Hamled) [inline specs/player_spec.rb:24]: This use of the `@debug` instance variable and special debug parameter for `draw_tiles()` is great!

It's very important for our testing that we have consistent behavior, otherwise we might get a test that passes _sometimes_, but other times it fails. And as you've identified, a major part of ensuring that consistent behavior is to avoid any kind of randomness (like with `shuffle!`). So providing a specific set of letters to populate the player's `@tiles` array with is a good thing.

However, the cost of this approach is that we're not testing the portion of the `draw_tiles` method that interacts with the `TileBag` instance, at least in these cases (there are some other cases that do run that code to test other things).

One alternative way to do the "special debug code for testing" approach is what we call "test mocking". With this approach we would create a fake `TileBag` instance (known as a "mock") which we would write so that it returns a specific sequence of tiles:

```ruby
class TileBagMock
  def initialize
    @tiles = ["P", "E", "R", "E", "Z", "A", "M", "M", "A"]
  end

  def draw_tiles(n)
    return @tiles.shift(n)
  end
end
```

This `TileBagMock` class does not have any randomness involved, but it presents the same "interface" to the `Player` class so we can swap it in by making the `@tile_bag` variable in our test code be an instance of `TileBagMock` instead of `TileBag`. The biggest advantage of this is that we get consistent behavior but we _haven't_ modified the code that we're testing (the `Player` class).

This technique is very common, and we'll see it a little bit when we get towards the end of the Rails curriculum. If you're curious though I would suggest Googling around for Mocking as there are several libraries designed to make this kind of thing easy -- Minitest even includes its own library for that.

File: lib/player.rb [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    def total_score
+      return @score_array.inject(0){|sum, x| sum + x}
+    end #total score end
+
+    def won?
+      if total_score > 100
+        return true
+      else
+        return false
+      end
```
Comment (Hamled) [inline lib/player.rb:46]: This `if` statement could be simplified to:
```ruby
return total_score > 100
```

File: lib/player.rb [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    end #won end
+
+    def highest_scoring_word_score
+      if @score_array  == []
+        raise ArgumentError.new("No words to compare")
+      else
+        index = @score_array.index(@score_array.max)
+        max_word = @plays[index]
+        max_score = @score_array.max
+        return max_word, max_score
```
Comment (Hamled) [inline lib/player.rb:56]: This method could be simpler if we don't keep separate `@plays` and `@score_array` arrays. Instead we could just have `@plays` and use the functionality from the `Scoring` class to determine the scores as-needed:

```ruby
def highest_scoring_word_score
  raise ArgumentError.new("No words to compare") if @plays.empty?

  best_word = Scoring.highscore_from(@plays)
  return best_word, Scoring.score(best_word)
end
```

File: lib/player.rb [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+      end
+    end #play word end
+
+    def delete_tiles_from_tiles_array(word)
+      word = word.upcase
+      word.split("").each do |letter|
+        if @tiles.include?(letter)
+          @tiles.slice!(@tiles.index(letter))
+        else
+          raise ArgumentError.new("You cannot play this word.")
```
Comment (Hamled) [inline lib/player.rb:32]: While this appears to be doing what we want, it's actually got some subtly incorrect behavior. Here's a test case that might expose what is going wrong:

```ruby
it "doesn't remove player's tiles if the word cannot be played" do
  jugador = Scrabble::Player.new("Morsa", true)
  jugador.draw_tiles(@tile_bag, ["C", "A", "C", "K", "L", "E", "S"])

  proc { jugador.play("CACKLED") }.must_raise ArgumentError
  jugador.play("CACKLES").must_equal 15
end
```

Try adding that to the test suite for `Player#play` and then figure out what changes are needed to make it pass.

File: lib/player.rb (hunk 6 above)
Comment (Hamled) [reply lib/player.rb:56]: In addition, the project requirements specify two methods, `highest_scoring_word` and `highest_word_score`. It looks like y'all have combined the functionality of those two methods into this method `highest_scoring_word_score`.

In the future we should strive to match the requirements specification as much as possible, so this method should be broken up into the two separate methods that each calculate one part of the two return values that this method is currently returning.
